### PR TITLE
Add AutoPauseSystem

### PR DESCRIPTION
**Pull request by nicolasfara**
Add `AutoPauseSystem`

### CODE REVIEW

File: demo/src/test/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystemTest.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  "An AutoPauseSystem" should {
+    "run" when {
+      "in an enabled state" in {
+        checkAllStates((playState, _) => AutoPauseSystem(playState))(
+          (State.Play, AnyValue, AnyValue, AnyValue),
+        )
+      }
+    }
+    "Pause the game" when {
```
Comment (giacomocavalieri) [inline demo/src/test/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystemTest.scala:21]:
```suggestion
    "pause the game" when {
```

File: demo/src/test/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystemTest.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      "the enery of the system is 0" in new AutoPauseSystemFixture {
```
Comment (giacomocavalieri) [inline demo/src/test/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystemTest.scala:22]:
```suggestion
      "the energy of the system is 0" in new AutoPauseSystemFixture {
```

File: demo/src/test/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystemTest.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        playState.gameState = State.Play
+        world hasAn entity withComponent Velocity(0.0, 0.0)
+        world hasAn entity withComponent Velocity(0.0, 0.0)
+        world.update(10)
+        playState.gameState shouldBe State.Pause
+      }
+    }
+    "do nothing" when {
+      "the enery of the system is not 0" in new AutoPauseSystemFixture {
```
Comment (giacomocavalieri) [inline demo/src/test/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystemTest.scala:31]:
```suggestion
      "the energy of the system is not 0" in new AutoPauseSystemFixture {
```

**Review by vitlinda — APPROVED** [review]

File: demo/src/main/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystem.scala [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package dev.atedeg.ecscalademo.systems
+
+import dev.atedeg.ecscala.given
+import dev.atedeg.ecscala.{ &:, CNil, Deletable, DeltaTime, Entity, IteratingSystem, View, World }
+import dev.atedeg.ecscalademo.given
+import dev.atedeg.ecscalademo.{ PlayState, State, Vector, Velocity }
+
+/**
+ * Pause the simulation when the system's energy is zero (all the ball has velocity = 0).
```
Comment (giacomocavalieri) [inline demo/src/main/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystem.scala:9]:
```suggestion
 * Pause the simulation when the system's energy is zero (all the balls have velocity = 0).
```

File: demo/src/main/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystem.scala [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+class AutoPauseSystem(private val playState: PlayState) extends IteratingSystem[Velocity &: CNil] {
+
+  override def shouldRun: Boolean = playState.gameState == State.Play
+
+  override def update(
+      entity: Entity,
+      components: Velocity &: CNil,
+  )(deltaTime: DeltaTime, world: World, view: View[Velocity &: CNil]): Deletable[Velocity &: CNil] = {
+    val systemEnergy = view.map(_._2.h).foldLeft(0.0)(_ + _.velocity.squaredNorm)
+    if (systemEnergy == 0.0) {
```
Comment (ndido98) [inline demo/src/main/scala/dev/atedeg/ecscalademo/systems/AutoPauseSystem.scala:20]: Use a fuzzy match (i.e. don't use `== 0` but `<= EPSILON`, I'd say a good epsilon is like 0.001)

**Review by ndido98 — CHANGES_REQUESTED** [review]

**Review by ndido98 — APPROVED** [review]

**Review by giacomocavalieri — APPROVED** [review]

**Pull request merged**